### PR TITLE
Maintenance week updates

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context

Update app according to [DataEng Application Maintenance document](https://mitlibraries.atlassian.net/wiki/spaces/IN/pages/3445948428/DataEng+Application+Maintenance) and fix logout method. 

As discussed, only a subset of our standard config and linting updates are applied for this iteration. Namely, the updates introduced are:
* Updates to the README
   * Add ‘Development’ section to describe commonly used commands and instructions for running the app
   * Add ‘Environment variables’ section to describe required and optional env vars.
* Update Pipfile
   * Add shortcut to run app
   * Update dependencies (e.g., general updates, addition of linting and stub packages
* Add Makefile
   *  Note: `ruff` has not been applied as a linter.
* Update Github PR template
* Add `pyproject.toml`
   * Include configs for `black`
   * Include configs for `mypy`
      * For `mypy` we've disabled several configs for this iteration. Overall, it may be worth considering whether to require strict typehinting for a Django app repo such as this (factors include complexity and not knowing how long this app will/should be supported).
         * `disallow_untyped_defs=false`: Setting this as `true` requires that all functions are typed (parameters and return values); there are a lot of methods that require further understanding before typing can be applied.
         * overriding error code `union-attr`  for `solenoid.elements.xml_handlers` (created Jira ticket [IN-979](https://mitlibraries.atlassian.net/browse/IN-979) in response
* Resolve linting errors
   * `black` formatting errors
   * `myy` minimal errors 

The issue with the logout method is described [here](https://forum.djangoproject.com/t/deprecation-of-get-method-for-logoutview/25533). The solution in commit 
https://github.com/MITLibraries/solenoid/pull/470/commits/988b9b1ef0ce71d7e459ff8110be8c5ec59925b2 replaces the logout method to use `POST` instead.

### How can a reviewer manually see the effects of these changes?
**Note:** The `staging` app connects to the Dev instance of MIT's Elements (pubdata-dev.mit.edu), so test-sending an email (with the configs indicated in the README) shouldn't be an issue!

1. Visit the Heroku app in staging: https://mitlibraries-solenoid-staging.herokuapp.com/
   * There was an issue with sending API requests to Elements through the 'Review app', so the app was instead manually deployed by pointing to this branch on Heroku.
2. Log in with the credentials shared with you on LastPass.
3. Run an import for Elements ID: 1902.
4. [Optional] Build and send test emails. Follow the instructions in [Solenoid: User Guide |  How to build emails](https://mitlibraries.atlassian.net/wiki/spaces/IN/pages/edit-v2/3655630876#How-to-build-emails)
   * Select the citations for "Zhang, Feng".
   * If you choose to send an email,  you can view the standard output on [logz.io](https://app.logz.io/#/dashboard/osd/discover/?_a=(columns:!(message),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'logzioCustomerIndex*',key:app,negate:!f,params:(query:mitlibraries-solenoid-staging),type:phrase),query:(match_phrase:(app:mitlibraries-solenoid-staging)))),index:'logzioCustomerIndex*',interval:auto,query:(language:lucene,query:''),sort:!())&_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-30m,to:now))&switchToAccountId=31815). The logs will look like this:
image

### Includes new or updated dependencies?
YES; Add `django-stubs`, `types-beautifulsoup4`, `types-requests`, and `ruff` (not actively used in linting workflow)

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/IN-958
- https://mitlibraries.atlassian.net/browse/IN-894

### Developer
- [x] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [ ] All related Jira tickets are linked in commit message(s)
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [x] The commit message is clear and follows our guidelines (not just this PR message)
- [x] There are appropriate tests covering any new functionality
- [x] The provided documentation is sufficient for understanding any new functionality introduced
- [x] Any manual tests have been performed or provided examples verified
- [x] New dependencies are appropriate or there were no changes

[IN-979]: https://mitlibraries.atlassian.net/browse/IN-979?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by jonavellecuerdo** [comment]
@ehanson8 @ghukill I made a small update to the README | Development section. See [latest commit](https://github.com/MITLibraries/solenoid/pull/470/commits/85be58b46c1f15a535e20ac8484aa75e06b8faf8). 

**Context:** Initially, I thought we'd be able to run the full Solenoid workflow via "Review app", but there appears to be some firewall configs set for the Dev instance of Symplectic Elements that--to our knowledge--only allows the IPs associated with the `staging` and `production` environments for the app.

**Review by ghukill — APPROVED** [review]
Approved!

The updates to the README, Makefile, Pipfile, etc., all feel like good, incremental changes to modernize this app according to our profiles.

I also think it was the right decision to hold off on full typing across the entire Django app, for the reasons you mentioned.

Probably out of scope of this PR, but this app clearly has some rough edges that make it tricky when considering how to update and maintain it.  For those reasons, I think this PR hits a nice middle ground of making incremental updates, improves our ability to maintain it, but doesn't over-commit to the current app and architecture.  As tickets are raised (and it sounds like some were created?), then we can use those to evaluate if undertaking those more invasive modernizations would be worth it, or if this app might need to be re-evaluated for how long, and how much, we want to support it.

File: README.md [hunk 1]
```diff
@@ -2,14 +2,114 @@
...
 [![Code Climate](https://codeclimate.com/github/MITLibraries/solenoid/badges/gpa.svg)](https://codeclimate.com/github/MITLibraries/solenoid)
 
-# README
+# solenoid
 
-Instructions for setting the system up for local development are in `docs/README.md`. The `system_architecture.xml` file is designed to be read by draw.io.
+Solenoid is a web application for sending email requests for publications to students and faculty. The application supports the work of MIT Libraries' Scholarly Communications and Collections Strategy (SCCS) department in the implementation of the university's open access (OA) policies. Given an author's ID on [MIT's instance of Symplectic Elements](https://pubdata.mit.edu/), Solenoid retrieves citations for publications that either (a) have not been deposited (to DSpace) or (b) have not been requested. Users can build emails using Solenoid, which provides a standard template for publication requests that can be revised through the application. Users have the option of either (a) sending the email to the Liaison or (b) copying the contents of the generated email and sending the email to the author directly.
```
Comment (ehanson8) [inline README.md:7]: Wonderful overview documentation, this is a great practice we're developing

**Review by ehanson8 — APPROVED** [review]
Agree with Graham about this being a good approach to linting for this app at the moment, excellent documentation as always

**Comment by jonavellecuerdo** [comment]
Thank you, both! I appreciate your flexibility and understanding as we gradually learn more about how Solenoid works!  🤓 Jira ticket [IN-978](https://mitlibraries.atlassian.net/browse/IN-978) has been created to track ideas for improvements that came up during maintenance and onwards. As @ghukill said, we can--at some point--review those tickets and decide how we want to proceed then.

**Pull request merged**